File: leetcode/arrays-hashing/049_group_anagrams.py
"""
Problem: Group Anagrams

Link: https://leetcode.com/problems/group-anagrams/

"""

from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = defaultdict(list)

        for word in strs:
            key = "".join(sorted(word))
            groups[key].append(word)

        return list(groups.values())


# Example
strs = ["eat", "tea", "tan", "ate", "nat", "bat"]

solution = Solution()

print(solution.groupAnagrams(strs))

File: leetcode/arrays-hashing/049_group_anagrams.php
<?php

/*
Problem: Group Anagrams

Link: https://leetcode.com/problems/group-anagrams/

*/

class Solution
{
    /**
     * @param array $strs
     * 
     * @return array
     */
    public function groupAnagrams(array $strs): array
    {
        $groups = [];

        foreach ($strs as $word) {

            $chars = str_split($word);
            sort($chars);

            $key = implode("", $chars);

            if (!isset($groups[$key])) {
                $groups[$key] = [];
            }

            $groups[$key][] = $word;
        }

        return array_values($groups);
    }
}


// Example
$strs = ["eat", "tea", "tan", "ate", "nat", "bat"];

$solution = new Solution();

print_r($solution->groupAnagrams($strs));